### DISCUSSION TITLE
Mass-flow rate distribution in meshed network

### DISCUSSION THREAD

**Discussion by MikhailSkl** [e2nIEE/pandapipes#494] [Q&A] [answered]
Hello everyone!

I'm trying to simulate a small and simple district heating network with one source and two loads (as heat exchangers) following a tutorial. Figure of the network and model is here:

![Untitled Diagram](https://user-images.githubusercontent.com/51205540/214298814-776a6bc7-9042-4a89-ab9c-20b346eeb2de.jpg)

```
import pandapipes as pp

############ Initialize network parameters ###############
pn_init_s = 8 # Initial nodal pressure in supply nodes [bar]
tfluid_init_s = 80 # Initial temperature in supply nodes [Deg C]

pn_init_r = 4 # Initial nodal pressure in return nodes [bar]
tfluid_init_r = 40 # Initial temperature in return nodes [C]
# Grid parameters
mdot_grid = 20 # Constant mass flow, which is injected by grid [kg/s]
P_grid_bar = 9  # Pressure of the external grid [bar]
T_supply_grid = 85  # Supply temperature of the external grid [degC]

# Ambient temperature

T_amb = 15 # Ambient temperature [degC]

# Heat consumers
Qdot_cons1 = 1500 # Heat consumption at Årsdale  [kW]
Qdot_cons2 = 1500 # Heat consumption at Svaneke  [kW]

############################# create empty net ###################
net = pp.create_empty_network(fluid ="water")

############################## create junctions #################
#supply
pp.create_junction(net, pn_bar=pn_init_s, tfluid_k =273.15 + tfluid_init_s, name='j1s')
pp.create_junction(net, pn_bar=pn_init_s, tfluid_k =273.15 + tfluid_init_s, name='j2s')
pp.create_junction(net, pn_bar=pn_init_s, tfluid_k =273.15 + tfluid_init_s, name='j3s')
#return
pp.create_junction(net, pn_bar=pn_init_r, tfluid_k =273.15 + tfluid_init_r, name='j1r')
pp.create_junction(net, pn_bar=pn_init_r, tfluid_k =273.15 + tfluid_init_r, name='j2r')
pp.create_junction(net, pn_bar=pn_init_r, tfluid_k =273.15 + tfluid_init_r, name='j3r')

############ Create circular pump flow with build-in external grid ##############    
pp.create_circ_pump_const_mass_flow(net, from_junction=net.junction['name'].tolist().index('j1s'), to_junction=net.junction['name'].tolist().index('j1r'), p_bar= P_grid_bar,
                                    mdot_kg_per_s = mdot_grid, t_k = 273.15 + T_supply_grid, type = 'pt' , name = 'ext_grid')

############### create heat exchangers #######################################
pp.create_heat_exchanger(net, from_junction=net.junction['name'].tolist().index('j2s'), to_junction=net.junction['name'].tolist().index('j2r'), diameter_m=200e-3, qext_w = Qdot_cons1*1000, name = 'h_ex_1')
pp.create_heat_exchanger(net, from_junction=net.junction['name'].tolist().index('j3s'), to_junction=net.junction['name'].tolist().index('j3r'), diameter_m=200e-3, qext_w = Qdot_cons2*1000, name = 'h_ex_2')

############### Create pipes ################################################
# supply pipes
pp.create_pipe_from_parameters(net, from_junction=net.junction['name'].tolist().index('j1s'), to_junction=net.junction['name'].tolist().index('j2s'),
                    length_km= 2.9, diameter_m=0.139, k_mm=0.01, sections=5, alpha_w_per_m2k=1.5,
					   text_k=273.15+T_amb, name='p1s')

pp.create_pipe_from_parameters(net, from_junction=net.junction['name'].tolist().index('j2s'), to_junction=net.junction['name'].tolist().index('j3s'),
                    length_km= 2.9, diameter_m=0.114, k_mm=0.01, sections=5, alpha_w_per_m2k=1.5,
					   text_k=273.15+T_amb, name='p2s')

### return pipes

pp.create_pipe_from_parameters(net, from_junction=net.junction['name'].tolist().index('j3r'), to_junction=net.junction['name'].tolist().index('j2r'),
                    length_km= 2.9, diameter_m=0.139, k_mm=0.01, sections=5, alpha_w_per_m2k=1.5,
					   text_k=273.15+T_amb, name='p1s')

pp.create_pipe_from_parameters(net, from_junction=net.junction['name'].tolist().index('j2r'), to_junction=net.junction['name'].tolist().index('j1r'),
                    length_km= 2.9, diameter_m=0.114, k_mm=0.01, sections=5, alpha_w_per_m2k=1.5,
					   text_k=273.15+T_amb, name='p2s')

    
####### Run a power flow ############
#pp.diagnostic(net)
pp.pipeflow(net, mode='all', transient=False, max_iter=1000, run_control=False, heat_transfer=True )

junctions = net.junction

pipes = net.pipe

res_junction = net.res_junction

res_pipe = net.res_pipe

res_circ_pump_mass = net.res_circ_pump_mass

res_heat_exchanger = net.res_heat_exchanger
```

I set a mass-flow for a source pump of 20 kg, and temperature difference of 45 deg C. That will give an input power around 4.167x20000x45 = 3 750 300 W or 3750 kW.

Next, I set two exchanger stations with 1500 kW of load each. I would assume that the same loads would get more or less equal mass flow distributions between them (around 10 kg/s each).  

However, the result I get, that almost all the mass flow (19.999 kg/s) goes to the fisrt heat echanger (HE1), and only 0.0001 kg/s goes to pipe PS2 and to HE2. Due to this the temperature difference in HE2 becomes inadequate. 

Is there any parameters that affect this mass flow rates distribution?

**Comment by jkisse** [comment]
Hi @MikhailSkl,

maybe the pressure drop at P2S and P2R is too high, so that the flow rate becomes very low. The problem should become smaller, if the diameters for these pipes are increased.

Apart from that, a [flow control component](https://pandapipes.readthedocs.io/en/v0.8.3/components/flow_control/flow_control_component.html) has been added to pandapipes as of Version 0.8. With this component, a mass flow set point between two junctions can be defined.  It could be well suited for your situation.

**Reply by MikhailSkl** [reply]
> Hi @MikhailSkl,
> 
> maybe the pressure drop at P2S and P2R is too high, so that the flow rate becomes very low. The problem should become smaller, if the diameters for these pipes are increased.
> 
> Apart from that, a [flow control component](https://pandapipes.readthedocs.io/en/v0.8.3/components/flow_control/flow_control_component.html) has been added to pandapipes as of Version 0.8. With this component, a mass flow set point between two junctions can be defined. It could be well suited for your situation.

it is very intresting, because the pressure drop in PS2 and PR2 is almost zero because here is almost no mass flow flowing. Pressure drop in PS1 is 2.13 bar when 20 kg/s flowing, so if the presure drop in PS3 and PR2 should be even less since less mass flow would flow there, so 9 bar should be enoguh for 4 pipes. I trided set pressure at source pump higher, does not help.

**Comment by dlohmeier** [comment, marked as answer]
Dear @MikhailSkl ,
in general, @jkisse is right, but I would like to shift the focus a little: The resistance rates of the two paths through the heat exchangers are extreme. Actually it would make sense if the flow through P2S and P2R would be 0, because there is a resistance of this pipe while there is no resistance in the heat exchanger HE1. So, if there is no pressure drop between J2S and J2R, any flow through P2S and P2R will create a higher pressure drop than that. So, I am not sure what exactly you would like to simulate, but here are some suggestions:
- Set the loss_coefficient of the heat_exchangers to some values higher than 0 --> if you want to simulate a regulator here, you can choose a high value for HE1 (that must compensate for the high resistance in P2S and P2R) and a lower one for HE2.
- Use the flow_control mentioned by @jkisse, which is a model of a perfect flow regulation valve with adaptable diameter
- Create your own flow regulation valve (actually, a new model is currently being implemented, but not planned to be released soon), e.g. by using a normal valve before the heat exchanger with its own loss coefficient that you could even control with a controller that checks a certain flow rate set point.

The suggestion to just increase the pipe diameters will probably not help much, as this does not change the fact that the resistance in HE1 remains 0.
Please note that in Version 0.8 of pandapipes, some arguments have changed wrt. the circulation pump model. You can easily find the changes in the creation function.

**Reply by MikhailSkl** [reply]
Thank you very much for  thorough explanation! Maybe it would be then logical to make 'loss_coeficient' parameter in Heat Exchanger component as mandatory? Also it would be nice to have a physical model and equations in the description of Heat Exchanger component.

**Reply by MikhailSkl** [reply]
It helped. I assigned loss_coficcient of 1000 for every HE, and the flows now are around 10 kg/s in each. However, now I'm facing a bit different issue :) When I calculate power extracted by the heat exchanger as 4.186 x 1000 x m_dot_from * (t_from - t_to), It does not match the assigned qext_w. Is there any thermal losses in the Heat Exchanger model?

**Reply by dlohmeier** [reply]
We will consider that. It's not an easy call to make between having a simple way to construct standard components and add them to your grid model, and hinting at data that might become necessary (we often have to balance out these aspects). About the documentation, you are certainly right... the main reason for missing documentation is usually missing time, but we will add this to our list (cf. #426)

**Reply by dlohmeier** [reply]
Well, that has something to do with the internal structure. If you calculate 4.186 x 1000 x m_dot_from * (t_from - t_to) for the second HE, the result should indeed be ~1500 kW. But for the first, you should note that t_to_k equals the temperature at the outlet junction, which has a mixed temperature from the two flows (one from HE1 and one from P2R). There is an internal node created (actually a little complex to extract this after the simulation), which has the resulting temperature. I think we should consider adding this internal temperature to the result table (cf. #495).

**Comment by dlohmeier** [comment]
I think I have to make an update here. The new controlling component is already included in the latest release, it's called flow_control and can be [created like any other component](https://github.com/e2nIEE/pandapipes/blob/0e74750fcef9265dd23120a66985f053b14f7513/src/pandapipes/create.py#L966).

**Reply by JonasPfeiffer123** [reply]
I would like to contribute to this discussion by sharing that recently, I have created several District Heating systems using pandapipes. My experience with the 'flow_control' component has been highly positive. It effectively regulates mass flow, particularly in conjunction with a controller. This combination allows for precise control of heat exchangers, ensuring a consistent return temperature.

**Reply by dlohmeier** [reply]
Thank you very much @JonasPfeiffer123 ! An important aspect to be mentioned about the flow controllers is that when using them to model all heat consumers, the generator unit (i.e. the circulation pump) type should be the one with constant pressure difference instead of constant mass flow rate, as the mass flow rate is then given by the flow controllers and a closed loop with mass flow controlling units at every connection of flow and return can lead to numeric issues.
